usecases/replica/coordinator.go

```diff
@@ -190,38 +213,86 @@ func (c *coordinator[T]) Pull(ctx context.Context,
+
+		// put the "backups/fallbacks" on the retry queue
+		for i := level; i < len(hosts); i++ {
+			hostRetryQueue <- hostRetry{
+				hosts[i],
+				backoff.WithContext(utils.NewExponentialBackoff(c.pullBackOffPreInitialInterval, c.pullBackOffMaxElapsedTime), ctx),
+			}
+		}
+
+		// kick off only level workers so that we avoid querying nodes unnecessarily
 		wg := sync.WaitGroup{}
-		wg.Add(len(candidates))
-		for i := range candidates { // Ask direct candidate first
-			idx := i
-			f := func() {
+		wg.Add(level)
+		for i := 0; i < level; i++ {
+			hostIndex := i
+			isFullReadWorker := hostIndex == 0 // first worker will perform the fullRead
+			workerFunc := func() {
 				defer wg.Done()
-				resp, err := op(ctx, candidates[idx], idx == 0)
-
-				// If node is not responding delegate request to another node
-				for err != nil {
-					if delegate, ok := <-candidatePool; ok {
-						resp, err = op(ctx, delegate, idx == 0)
-					} else {
-						break
+				// each worker will first try its corresponding host (eg worker0 tries hosts[0],
+				// worker1 tries hosts[1], etc). We want the fullRead to be tried on hosts[0]
+				// because that will be the direct candidate (if a direct candidate was provided),
+				// if we only used the retry queue then we would not have the guarantee that the
+				// fullRead will be tried on hosts[0] first.
+				resp, err := op(ctx, hosts[hostIndex], isFullReadWorker)
+				// TODO how to know if/handle op erroring retryable vs not
+				// TODO have increasing timeout passed into each op (eg 1s, 2s, 4s, 8s, 16s, 32s, with some max) similar to backoff? future PR? or should we just set timeout once per worker in Pull?
+				if err == nil {
+					replyCh <- _Result[T]{resp, err}
+					return
+				}
+				// this host failed op on the first try, put it on the retry queue
+				hostRetryQueue <- hostRetry{
+					hosts[hostIndex],
+					backoff.WithContext(utils.NewExponentialBackoff(c.pullBackOffPreInitialInterval, c.pullBackOffMaxElapsedTime), ctx),
+				}
+
+				// let's fallback to the backups in the retry queue
+				for hr := range hostRetryQueue {
+					resp, err := op(ctx, hr.host, isFullReadWorker)
+					if err == nil {
+						replyCh <- _Result[T]{resp, err}
+						return
 					}
+					nextBackOff := hr.currentBackOff.NextBackOff()
+					if nextBackOff == backoff.Stop {
+						// this host has run out of retries, send the result and note that
+						// we have the worker exit here with the assumption that once we've reached
+						// this many failures for this host, we've tried all other hosts enough
+						// that we're not going to reach level successes
+						replyCh <- _Result[T]{resp, err}
+						return
+					}
+
+					timer := time.NewTimer(nextBackOff)
+					select {
+					case <-ctx.Done():
+						timer.Stop()
+						replyCh <- _Result[T]{resp, err}
+						return
+					case <-timer.C:
+						hostRetryQueue <- hostRetry{hr.host, hr.currentBackOff}
+					}
+					timer.Stop()
```

Comment:
**suggestion:** isn't this gonna be redundant with every worker, wdyt if we have a routine just for pulling from the queue running before the worker 🤔

Reply:
e.g. 

```go
hostRetryQueue := make(chan hostRetry, len(hosts))
enterrors.GoWrapper(func() {
	c.queueWorker(ctx, hostRetryQueue, op, replyCh)
}, c.log)
```

```go
// hostRetry tracks how long we should wait to retry this host again
type hostRetry struct {
	host             string
	currentBackOff   backoff.BackOff
	isFullReadWorker bool
}

func (c *coordinator[T]) queueWorker(ctx context.Context, hostRetryQueue chan hostRetry, op readOp[T], replyCh chan _Result[T]) {
	// let's fallback to the backups in the retry queue
	for hr := range hostRetryQueue {
		resp, err := op(ctx, hr.host, hr.isFullReadWorker)
		if err == nil {
			replyCh <- _Result[T]{resp, err}
			return
		}
		nextBackOff := hr.currentBackOff.NextBackOff()
		if nextBackOff == backoff.Stop {
			// this host has run out of retries, send the result and note that
			// we have the worker exit here with the assumption that once we've reached
			// this many failures for this host, we've tried all other hosts enough
			// that we're not going to reach level successes
			replyCh <- _Result[T]{resp, err}
			return
		}

		timer := time.NewTimer(nextBackOff)
		select {
		case <-ctx.Done():
			timer.Stop()
			replyCh <- _Result[T]{resp, err}
			return
		case <-timer.C:
			hostRetryQueue <- hostRetry{hr.host, hr.currentBackOff, hr.isFullReadWorker}
		}
		timer.Stop()
	}
}
```

Reply:
I'd like to explore this further, this is code cleanliness and not correctness/perf related, right? If so, mind if I explore in a follow up PR?